src/libsyntax/parse/parser.rs

```diff
@@ -4186,11 +4175,25 @@ impl<'a> Parser<'a> {
                                         mode: BoundParsingMode)
                                         -> PResult<'a, TyParamBounds>
     {
-        if !self.eat(&token::Colon) {
-            Ok(P::new())
-        } else {
-            self.parse_ty_param_bounds(mode)
+        Ok(self.parse_opt_ty_param_bounds(&token::Colon, mode)?.unwrap_or_else(P::new))
+    }
+
+    fn parse_opt_ty_param_bounds(&mut self,
+                                 intoducing_token: &token::Token,
+                                 mode: BoundParsingMode)
+                                 -> PResult<'a, Option<TyParamBounds>> {
+        if !self.eat(intoducing_token) {
+            return Ok(None);
         }
+
+        let bounds = self.parse_ty_param_bounds(mode)?;
```

Comment:
@matklad `parser.eat(&token::Ident(keywords::Impl.ident()))` should do the trick.